File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        area = 0
        leftHeight, rightHeight = [ 0 for _ in range(len(height))] , [ 0 for _ in range(len(height))]
        
        for i in range(0,len(height)-1):
            leftHeight[i+1] = max(leftHeight[i],height[i])
        print(leftHeight)
        
        for i in range(len(height)-1,0,-1):
            rightHeight[i-1] = max(rightHeight[i],height[i])
        print(rightHeight)
        for i in range(1,len(height)-1):
            area += min(leftHeight[i],rightHeight[i]) - height[i] if min(leftHeight[i],rightHeight[i]) - height[i] > 0 else 0
        return area
        
        

File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int[] left = new int[height.length] ; 
        int[] right = new int[height.length] ; 
        int leftmax = 0;
        int rightmax = 0;
        int ans = 0 ;
        for (int i = 0 ; i < height.length ; i ++){
            left[i] = leftmax;
            leftmax = Math.max(leftmax,height[i]);
        }
        for (int i = height.length - 1 ; i >= 0 ; i --){
            right[i] = rightmax;
            rightmax = Math.max(rightmax,height[i]);
        }
        for (int i = 0 ; i < height.length ; i ++){
            ans += Math.max(0,Math.min(right[i],left[i]) - height[i]);
        }
        return ans ;
        
    }
}